README.rst
django-stagesetting 0.5.0
=========================

An application for managing site configuration through normal `Django`_ forms,
and thus through the `admin site`_.

.. |travis_stable| image:: https://travis-ci.org/kezabelle/django-stagesetting.svg?branch=0.5.0
  :target: https://travis-ci.org/kezabelle/django-stagesetting

.. |travis_master| image:: https://travis-ci.org/kezabelle/django-stagesetting.svg?branch=master
  :target: https://travis-ci.org/kezabelle/django-stagesetting

==============  ======
Release         Status
==============  ======
stable (0.5.0)  |travis_stable|
master          |travis_master|
==============  ======

Pre-requisites
--------------

The following versions are tested:

* Python 2.7, or 3.3+
* `Django`_ 1.7+

Installation
------------

First up, you need to install it  (via `pip`_ as usual)::

    pip install django-stagesetting==0.5.0

Once that's downloaded, add the package to your ``INSTALLED_APPS``
in your settings::

    INSTALLED_APPS = (
        # ...
        'stagesetting',
        # ...
    )

do a migrate::

    python manage.py migrate stagesetting

Add a ``STAGESETTINGS`` dictionary to your project's settings::

    STAGESETTINGS = {
        'SETTING_NAME': '...',
        'ANOTHER_SETTING_NAME': '...',
    }

The setting collection name is the dictionary key, so must be unique.

Writing settings
^^^^^^^^^^^^^^^^
Settings may be created in a number of ways, the simplest of which is to
provide a ``dictionary`` as the value::

    STAGESETTINGS = {
        'MY_SETTING': {
            'an_example-datetime': datetime.today(),
            'a_date': date.today(),
            'time_now': time(4, 23),
            'boolean_field': False,
            'plain_text': 'char field',
            'decimal': Decimal('3.25'),
            'float': 2.3,
        }
    }

where possible, this will auto-generate a Form class for you, choosing sensible
defaults for the field variants where possible.

The other option is for the value to be a ``list`` or a ``tuple``, where
the *first item* represents a form (either a ``dictionary`` as above, **OR**
the ``dotted.path.to.a.Form.Class`` if you need custom validation) and the
*second, optional item* is the default data. The following should all be valid::

    STAGESETTINGS = {
        'AUTO_GENERATED': [{
            'datetime': datetime.today(),
        }],
        'IMPORT_A_FORM': ['myapp.forms.MyForm'],
        'IMPORT_WITH_DEFAULT': ['myapp.forms.MyForm', {'default': 'data'}],
        'AUTO_GENERATED_WITH_OTHER_DEFAULTS': [{
            'datetime': datetime.today(),
        }, {'default': 'data'}],
    }

A simple configuration form (for the ``dotted.path.Format``) might look like::

    from django.core.exceptions import ValidationError
    from django.forms import Form, DateField

    class DateForm(Form):
        start = DateField()
        end = DateField()

        def clean(self):
            cd = self.cleaned_data
            if 'start' in cd and 'end' in cd and cd['start'] > cd['end']:
                raise ValidationError("Start date cannot be after end date")
            return cd

As you can see, it really is just a normal `Form`_. Internally, this form's
``cleaned_data`` will be converted into `JSON`_ before being saved to the
database.
It will get re-converted to proper Python values when pulled out
of the database, by going through the given `Form`_ class's validation again,
including converting to rich values like model instances.


Python types which can be detected
**********************************

When detecting a dictionary as the value and auto-generating a form, the
following translations will be applied:

- ``None`` becomes `NullBooleanField`_
- ``datetime.datetime`` becomes `DateTimeField`_
- ``datetime.date`` becomes `DateField`_
- ``datetime.time`` becomes `TimeField`_
- ``datetime.timedelta`` becomes `DurationField`_
- ``decimal.Decimal`` becomes `DecimalField`_
- ``float`` becomes `FloatField`_
- ``True`` or ``False`` become `BooleanField`_
- ``int`` becomes `IntegerField`_
- ``uuid.UUID`` becomes `UUIDField`_ or `RegexField`_, depending on the `Django`_ version
- ``list`` and ``tuple`` become `MultipleChoiceField`_
- ``collections.OrderedDict``, ``set``, ``frozenset``, and ``dict`` become `ChoiceField`_
- ``models.Model`` instances become `ModelChoiceField`_
- ``models.QuerySet`` becomes `ModelMultipleChoiceField`_
- strings become one of the following, depending on what checks they pass:

  - `GenericIPAddressField`_
  - `URLField`_
  - `EmailField`_
  - `SlugField`_
  - `CharField`_
  - `IntegerField`_
  - `DecimalField`_
  - `UUIDField`_ (or `RegexField`_, depending on the `Django`_ version)
  - `DateTimeField`_
  - `DateField`_
  - `TimeField`_
- Some strings are **really** special, and will instead turn into one of the following:

  - if the string == ``STATIC_URL`` or ``STATICFILES_STORAGE`` the field will be
    a `ChoiceField`_ whose options are all the files found by the
    project's ``STATICFILES_FINDERS``.
  - if the string == ``MEDIA_URL``or ``DEFAULT_FILE_STORAGE`` the field will be
    a `ChoiceField`_ whose options are all the files found by
    ``DEFAULT_FILE_STORAGE``.
  - if the string *starts with* ``STATIC_URL`` it will be the same as using
    the ``STATIC_URL`` generated field, but is a regular expression for filtering
    to only certain files (i.e. ``/static/.*\.css$`` would find only css files)
  - if the string *starts with* ``MEDIA_URL`` it will be the same as above,
    but for files found in ``DEFAULT_FILE_STORAGE``.
  - if a string looks like it contains HTML, it will try to use `django-bleach`_
    for sanitisation, and one of `django-ckeditor`_, `django-tinymce`_,
    `django-markdown`_, `django-pagedown`_, or `django-epiceditor`_ for an
    appropriate widget.

Usage in code
-------------

The best way to access the settings in your views is to include
``stagesetting.middleware.ApplyRuntimeSettings`` in your ``MIDDLEWARE_CLASSES``
which will ensure there is a ``request.stagesettings`` variable which can be
used like so::

    def myview(request):
        how_many_form_data = request.stagesetting.LIST_PER_PAGE
        allow_empty_form_data = request.stagesetting['ALLOW_EMPTY']

each setting will be a dictionary of the `Form`_ values, either the default ones
or those changed in the database.

Usage in templates
------------------

If you've already got ``request`` in your template, obviously you can continue
to use ``request.stagesettings`` if the middleware is wired up.

If you don't have request, or you're not using the middleware,
``stagesetting.context_processors.runtime_settings`` provides a ``STAGESETTING``
template variable which contains the exact same data.

Finally, if not using the middleware nor the context processor, there is a
template tag available as a last resort. It's usage is::

    {% load stagesetting %}
    {% stagesetting as NEW_CONTEXT_VARIABLE %}
    {{ NEW_CONTEXT_VARIABLE.SETTING_NAME.fieldname }}


Usage outside of a request
--------------------------

If you don't have the middleware, or are in a part of the code which doesn't
have a ``request``, you can use the wrapper object directly::

    from stagesetting.models import RuntimeSettingWrapper
    def my_signal_handler(sender, instance, **kwargs):
        live_settings = RuntimeSettingWrapper()
        data = live_settings.LIST_PER_PAGE

Try to keep a single ``RuntimeSettingWrapper`` around for as long as possible,
rather than creating a new instance everywhere, as the object must fetch
the available settings from the database the first time it needs them. It
caches them for it's lifetime thereafter.

Alternatives
------------

Other apps I know of that achieve similar things, or overlap in some obvious
way. I won't judge you for using them, and I can't promise this is better.
To the victor, the spoils of maintenance!

- `django-constance`_ is similar

  - uses ``pickle`` to store an arbitrary python value; ``stagesetting`` only
    stores stuff it can put into JSON and relies on `Django`_ `Forms`_ to inflate
    the JSON back into python values.
  - Has both database and redis backends; ``stagesetting`` only supports
    the database, though it will only do one query most of the time.
    
- `django-dynamic-preferences`_ by the look of it.
- `django-solo`_ as well.

If you think GitHub popularity is an indication of usage and battle-tested 
production-readiness, then any of the above are certainly worth considering, 
being much more noticed than this, my attempt.

.. _Django: https://docs.djangoproject.com/en/stable/
.. _admin site: https://docs.djangoproject.com/en/stable/ref/contrib/admin/
.. _contrib.admin: https://docs.djangoproject.com/en/stable/ref/contrib/admin/
.. _Form: https://docs.djangoproject.com/en/stable/topics/forms/
.. _Forms: https://docs.djangoproject.com/en/stable/topics/forms/
.. _JSON: http://json.org/
.. _pip: https://pip.pypa.io/en/stable/
.. _pytest: http://pytest.org/latest/
.. _BooleanField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#booleanfield
.. _CharField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#charfield
.. _RegexField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#regexfield
.. _ChoiceField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#choicefield
.. _DateField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#datefield
.. _DateTimeField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#datetimefield
.. _DecimalField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#decimalfield
.. _DurationField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#durationfield
.. _EmailField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#emailfield
.. _FloatField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#floatfield
.. _GenericIPAddressField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#genericipaddressfield
.. _IntegerField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#integerfield
.. _ModelChoiceField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#modelchoicefield
.. _ModelMultipleChoiceField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#modelmultiplechoicefield
.. _MultipleChoiceField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#multiplechoicefield
.. _NullBooleanField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#nullbooleanfield
.. _SlugField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#slugfield
.. _TimeField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#timefield
.. _URLField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#urlfield
.. _UUIDField: https://docs.djangoproject.com/en/stable/ref/forms/fields/#uuidfield
.. _django-bleach: https://bitbucket.org/tim_heap/django-bleach
.. _django-ckeditor: https://github.com/django-ckeditor/django-ckeditor
.. _django-tinymce: https://github.com/aljosa/django-tinymce
.. _django-markdown: https://github.com/klen/django_markdown
.. _django-pagedown: https://github.com/timmyomahony/django-pagedown
.. _django-epiceditor: https://github.com/barraq/django-epiceditor
.. _django-constance: https://github.com/jezdez/django-constance
.. _django-dynamic-preferences: https://github.com/EliotBerriot/django-dynamic-preferences
.. _django-solo: https://github.com/lazybird/django-solo


CHANGELOG
Change log
----------
0.5.0
^^^^^^
* Minor changes to allow for easy subclassing or replacement of the middleware,
  context processor etc.
* Updated test matrix to cover more versions of Django (1.9+)


0.4.2
^^^^^^

* Admin now displays pretty names for the setting keys.
* ``StaticFilesChoiceField`` and ``DefaultStorageFilesChoiceField`` now both use an
  LRU cache to avoid having to enumerate the storage backends/finders repeatedly.


0.4.1
^^^^^^

* When generating a form from an ``OrderedDict``, field ordering is now preserved.
* When a form's values are saved into the database and become stale (because the
  underlying form/defaults are added to), the new defaults are transparently
  mapped onto the value when it's made available via the middleware,
  context processor or template tag.


0.4.0
^^^^^^

* Introduced the ``stagesetting`` template tag.
* Settings are no longer automatically created on app ``ready()``, instead
  the defaults are lazily converted when requested via a ``RuntimeSettingWrapper``
* Added ``StaticFilesChoiceField`` and ``DefaultStorageFilesChoiceField``,
  which allow for selecting a file that already exists in the given storage.
* When auto-generating a form, the ``STATIC_URL`` and ``MEDIA_URL`` settings
  are treated as special, and turned into the aforementioned fields.
* Providing an incomplete defaults dictionary in the second parameter of a
  given setting's config will now show *Info* messages to indicate what's
  missing.
* When auto-generating a form from a dictionary, HTML-like values will try
  to use one of `django-ckeditor`_, `django-tinymce`_, `django-markdown`_,
  `django-pagedown`_, or `django-epiceditor`_ for a widget instead of a normal
  ``textarea``.
* Added support for using `django-bleach`_ when encountering HTML-like values
  in an autogenerated form.
* Added initial support for `djangorestframework`_


0.3.2
^^^^^^

* Fixed error introduced in ``0.3.1`` when only providing a dictionary, because
  it turns out it wasn't being covered by tests.

0.3.1
^^^^^^

* Fixed issue where providing a dictionary wasn't treating the values as
  implicit defaults to be created into the database.


0.3.0
^^^^^^

* Added ability to auto-generate forms for configuration values which are
  dictionaries.

0.2.0
^^^^^^

* Initial release.

.. _djangorestframework: https://github.com/tomchristie/django-rest-framework
.. _django-bleach: https://bitbucket.org/tim_heap/django-bleach
.. _django-ckeditor: https://github.com/django-ckeditor/django-ckeditor
.. _django-tinymce: https://github.com/aljosa/django-tinymce
.. _django-markdown: https://github.com/klen/django_markdown
.. _django-pagedown: https://github.com/timmyomahony/django-pagedown
.. _django-epiceditor: https://github.com/barraq/django-epiceditor


demo_project.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import absolute_import
from __future__ import unicode_literals
from __future__ import print_function
from datetime import datetime, date, time
import os
import sys
sys.dont_write_bytecode = True
MISSING_DEPENDENCY = False
try:
    from django.conf import settings
except ImportError:
    sys.stdout.write("You'll need to `pip install Django>=1.7` to run "
                     "this demo\n")
    MISSING_DEPENDENCY = True

try:
    import debug_toolbar
except ImportError:
    sys.stdout.write("You'll need to `pip install django-debug-toolbar` to "
                     "run this demo\n")
    MISSING_DEPENDENCY = True
try:
    import rest_framework
except ImportError:
    sys.stdout.write("You'll need to `pip install djangorestframework>=3.2` "
                     "to run this demo\n")
    MISSING_DEPENDENCY = True

if MISSING_DEPENDENCY:
    sys.exit(1)


DEBUG = os.environ.get('DEBUG', 'on') == 'on'
SECRET_KEY = os.environ.get('SECRET_KEY', 'TESTTESTTESTTESTTESTTESTTESTTEST')
ALLOWED_HOSTS = os.environ.get('ALLOWED_HOSTS', 'localhost,testserver').split(',')

BASE_DIR = os.path.abspath(os.path.dirname(os.path.abspath(__file__)))

def user():
    from django.contrib.auth import get_user_model
    return get_user_model()()

def user_qs():
    from django.contrib.auth import get_user_model
    return get_user_model().objects.all()

settings.configure(
    DEBUG=DEBUG,
    SECRET_KEY=SECRET_KEY,
    ALLOWED_HOSTS=ALLOWED_HOSTS,
    SITE_ID=1,
    ROOT_URLCONF='test_urls',  # or __name__ to use local ones ...
    MIDDLEWARE_CLASSES=(
        'django.middleware.common.CommonMiddleware',
        'django.middleware.csrf.CsrfViewMiddleware',
        'django.middleware.clickjacking.XFrameOptionsMiddleware',
        'django.contrib.messages.middleware.MessageMiddleware',
        'django.contrib.sessions.middleware.SessionMiddleware',
        'django.contrib.auth.middleware.AuthenticationMiddleware',
        'stagesetting.middleware.ApplyRuntimeSettings',
    ),
    DATABASES={
        'default': {
            'ENGINE': 'django.db.backends.sqlite3',
            'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
        }
    },
    TEMPLATE_CONTEXT_PROCESSORS=(
        'django.contrib.messages.context_processors.messages',
        'django.contrib.auth.context_processors.auth',
        'stagesetting.context_processors.runtime_settings',
    ),
    INSTALLED_APPS=(
        'django.contrib.contenttypes',
        'django.contrib.messages',
        'django.contrib.sites',
        'django.contrib.auth',
        'django.contrib.staticfiles',
        'django.contrib.admin',
        'stagesetting',
        'test_app',
        'debug_toolbar',
        'rest_framework',
        # 'rest_framework_swagger',
    ),
    BLEACH_ALLOWED_TAGS=['p', 'h3', 'h4', 'em', 'strong', 'a', 'ul', 'ol', 'li', 'blockquote'],
    BLEACH_ALLOWED_ATTRIBUTES=['href', 'title', 'name'],
    BLEACH_STRIP_TAGS=True,
    STATIC_ROOT=os.path.join(BASE_DIR, 'test_collectstatic'),
    MEDIA_ROOT=os.path.join(BASE_DIR, 'test_app'),
    STATIC_URL='/__static__/',
    MEDIA_URL='/__media__/',
    MESSAGE_STORAGE='django.contrib.messages.storage.cookie.CookieStorage',
    SESSION_ENGINE='django.contrib.sessions.backends.signed_cookies',
    SESSION_COOKIE_HTTPONLY=True,
    STAGESETTINGS={
        'LIST_PER_PAGE': ['test_app.forms.ListPerPageForm'],
        'DATES': ['test_app.forms.DateForm', {'start': date.today()}],
        'DATETIMES': ['test_app.forms.DatetimeForm'],
        'USERS': ['test_app.forms.ModelChoicesForm'],
        'GENERATED':  {
            'int': 1,
            'email': 'a@b.com',
            'url': 'https://news.bbc.co.uk/',
            'model': user,
            'queryset': user_qs,
            'datetime': datetime.today(),
            'date': date.today(),
            'time': time(4, 23),
            'boolean': False,
            'nullboolean': None,
            'ip': '127.0.0.1',
            'slug': 'test-test',
            'text': 'char field',
        },
        'SPECIAL_STUFF':  [{
            'static_selection': '/__static__/',
            'static_storage': 'django.contrib.staticfiles.storage.StaticFilesStorage',
            'partial_static':  r'^/__static__/admin/(.+)\.css|\.txt$',
            'media_storage': 'django.core.files.storage.FileSystemStorage',
            'media_root': '/__media__/',
            'partial_media': r'^/__media__/(.+)\.txt$',
            'html': '<b>test</b>',
        }, {
            'static_selection': 'admin/css/login.css',
            'partial_static': 'admin/js/LICENSE-JQUERY.txt',
        }]
    },
)

from django.core.wsgi import get_wsgi_application
application = get_wsgi_application()


if __name__ == "__main__":
    from django.core.management import execute_from_command_line
    execute_from_command_line(sys.argv)
